src/client/Microsoft.Identity.Client/Cache/CacheSessionManager.cs

```diff
@@ -84,57 +84,67 @@ public async Task<IEnumerable<IAccount>> GetAccountsAsync()
+                if (!_cacheRefreshedForRead)
                 {
-                    TokenType = cacheEventType
-                };
+                    string telemetryId = _requestParams.RequestContext.CorrelationId.AsMatsCorrelationId();
+                    var cacheEvent = new CacheEvent(CacheEvent.TokenCacheLookup, telemetryId)
+                    {
+                        TokenType = cacheEventType
+                    };
 
-                await TokenCacheInternal.Semaphore.WaitAsync().ConfigureAwait(false);
+                    await TokenCacheInternal.Semaphore.WaitAsync().ConfigureAwait(false);
 
-                try
-                {
-                    if (!_cacheRefreshedForRead) // double check locking
+                    try
                     {
-                        using (_requestParams.RequestContext.CreateTelemetryHelper(cacheEvent))
+                        if (!_cacheRefreshedForRead) // double check locking
                         {
-                            string key = SuggestedWebCacheKeyFactory.GetKeyFromRequest(_requestParams);
-                            try
+                            using (_requestParams.RequestContext.CreateTelemetryHelper(cacheEvent))
                             {
-                                var args = new TokenCacheNotificationArgs(
-                                   TokenCacheInternal,
-                                   _requestParams.ClientId,
-                                   _requestParams.Account,
-                                   hasStateChanged: false,
-                                   TokenCacheInternal.IsApplicationCache,
-                                   hasTokens: TokenCacheInternal.HasTokensNoLocks(),
-                                   suggestedCacheKey: key);
-                                await TokenCacheInternal.OnBeforeAccessAsync(args).ConfigureAwait(false);
+                                string key = SuggestedWebCacheKeyFactory.GetKeyFromRequest(_requestParams);
+
+                                try
+                                {
+                                    var args = new TokenCacheNotificationArgs(
+                                       TokenCacheInternal,
+                                       _requestParams.ClientId,
+                                       _requestParams.Account,
+                                       hasStateChanged: false,
+                                       TokenCacheInternal.IsApplicationCache,
+                                       hasTokens: TokenCacheInternal.HasTokensNoLocks(),
+                                       suggestedCacheKey: key);
+                                    await TokenCacheInternal.OnBeforeAccessAsync(args).ConfigureAwait(false);
+                                }
+                                finally
+                                {
+                                    if (TokenCacheInternal.IsTokenCacheSerialized())
```

Comment:
I don't think this value will change, no need to check twice.

Reply:
Done.